saleor/warehouse/models.py

```diff
@@ -25,8 +30,66 @@ def for_country(self, country: str):
             .order_by("pk")
         )
 
+    def applicable_for_click_and_collect_no_quantity_check(
+        self, lines: QuerySet[CheckoutLine], country: str
+    ):
+
+        warehouse_cc_option_enum = WarehouseClickAndCollectOption
+        stocks_qs = Stock.objects.filter(
+            product_variant__id__in=lines.values("variant_id"),
+        ).select_related("product_variant")
+
+        warehouses_qs = (
+            self.for_country(country)
+            .prefetch_related(Prefetch("stock_set", queryset=stocks_qs))
+            .filter(stock__in=stocks_qs)
+            .annotate(stock_num=Count("stock__id"))
+            .filter(
+                Q(stock_num=lines.count())
+                & Q(click_and_collect_option=warehouse_cc_option_enum.LOCAL_STOCK)
+                | Q(click_and_collect_option=warehouse_cc_option_enum.ALL_WAREHOUSES)
+            )
+        )
+        return warehouses_qs
+
+    def applicable_for_click_and_collect(
+        self, lines: QuerySet[CheckoutLine], country: str
+    ):
+        warehouse_cc_option_enum = WarehouseClickAndCollectOption
+        lines_quantity = (
+            lines.filter(variant_id=OuterRef("product_variant_id"))
+            .annotate(prod_sum=Sum("quantity"))
+            .values_list("prod_sum")
+        )
+
+        stocks_qs = (
+            Stock.objects.annotate_available_quantity()
+            .annotate(line_quantity=F("available_quantity") - Subquery(lines_quantity))
+            .order_by("line_quantity")
+            .filter(
+                product_variant__id__in=lines.values("variant_id"),
+                line_quantity__gte=0,
+            )
+            .select_related("product_variant")
+        )
+
+        warehouses_qs = (
```

Comment:
This query is identical to the query in `applicable_for_click_and_collect_no_quantity_check`. Maybe extract it to a function? Could be a "private" function, only used locally in this queryset class.

Reply:
Done.